research/deep_speech/decoder.py

```diff
@@ -18,189 +18,78 @@
 from __future__ import division
 from __future__ import print_function
 
+import itertools
+
 from nltk.metrics import distance
-from six.moves import xrange
-import tensorflow as tf
+import numpy as np
 
 
 class DeepSpeechDecoder(object):
-  """Basic decoder class from which all other decoders inherit.
-
-  Implements several helper functions. Subclasses should implement the decode()
-  method.
-  """
+  """Greedy decoder implementation for Deep Speech model."""
 
-  def __init__(self, labels, blank_index=28, space_index=27):
+  def __init__(self, labels, blank_index=28):
     """Decoder initialization.
 
     Arguments:
-      labels (string): mapping from integers to characters.
-      blank_index (int, optional): index for the blank '_' character.
-        Defaults to 0.
-      space_index (int, optional): index for the space ' ' character.
+      labels: a string specifying the speech labels for the decoder to use.
+      blank_index: an integer specifying index for the blank character.
```

Comment:
'_' appears as the last element of our vocabulary as a pure placeholder character for blank output, so this is not really meant for outputting an underscore:) To work with CTC loss, it expects that we manually insert this placeholder into our vocabulary. In the last fully connected layer, the network will learn to output a blank symbol if there is no enough meaningful features at a time frame. For example, if I utter something like 'hello' and with the utterance may contain some disconnections, this will be captured by the network as an output form like 'h-ell-o'. The special '-' sign will be trimmed by CTC loss function when computing the actual loss.